### ISSUE TITLE
Make the Search service return more relevant results

### ISSUE THREAD

**Issue by SCdF** [medic/medic-webapp#2257]
Right now the free-text portion of the search service works querying a view that has an `emit` for every "token" of every document (that we care about, i.e. all Contacts). If a token is deemed to be a string it's broken up on space and emitted N times for each of _those_ tokens.

It is done this way because we do not impose a particular schema, so there is no "white list" of fields that we should be searching.

The downside of this is that it often isn't amazing at relevance. It returns _everything_ associated with your particular query.

This is particularly obvious that we now use the Search service in dropdowns:

![image](https://cloud.githubusercontent.com/assets/583851/14918627/e51f6d76-0e1c-11e6-8625-2c27d25b5fbd.png)

He's there, just buried:

![image](https://cloud.githubusercontent.com/assets/583851/14918643/f2f4363e-0e1c-11e6-8cdc-17c69c9bfedf.png)

**We need to work out how to make Search more relevant.**

# Options

## Weighted tokens

One idea is to make the view "weighted". You would change the view to, for every token, `emit([doc._id, token], 1);`. We would then group by the first key (the document id) and reduce by the value (a 1 for each result).

## Or something else?

However, I'm not convinced that weighted tokens will necessarily work for us in this case. That algorithm gives preference to documents that contain that value _more_, but that isn't necessarily going to lend itself to more relevance. In the above example all of those probably contain the same number of "victor" tokens.

There is currently undocumented sneaky hax in there, so you can do this:

![image](https://cloud.githubusercontent.com/assets/583851/14918892/082d08a4-0e1e-11e6-9a8f-83bd2d13683e.png)

I wonder if we need to think more in this direction?

**Comment by SCdF** [comment]
cc. @garethbowen

**Comment by estellecomment** [comment]
How important is it to get super-search? Maybe just indexing contact names would meet user expectations. Search for "Victor", get people and places named “XXXvictorYYY”, done.

What are user stories for search queries that are not contact names? 
Search in the notes, to find a person by place, because you described where they lived in the notes? 
Search by phone number? 
Searching by parent place is covered with the filters (select your branch/clinic and you get a subset).
@diannakane, @abbyad, thoughts ?

I vote for implementing a fast simple search, and waiting to see if anyone complains (and/or running user studies in a couple months to discover if there's a pain point, because people might not realize they have pain points there)

And/or, add a separate search page for advanced search, that's super slow and crashes your browser if you're not on macbook pro, but hey, it's advanced!

**Comment by abbyad** [comment]
From what I remember, the current behaviour is by design because we wanted all related people to show up. This is also the case for the search box in the Contact tab filter bar. It would be a good time for the design team to reevaluate this.

cc @diannakane

**Comment by estellecomment** [comment]
We could also search for all names within your parent places and contact persons, but only `name` field.

**Comment by SCdF** [comment]
> And/or, add a separate search page for advanced search, that's super slow and crashes your browser if you're not on macbook pro, but hey, it's advanced!

We're talking about the search _service_ here, not any particular search page. Searches can be a whole page affair like the Contacts page or the Reports page, or they can be inside select boxes, or <insert thing we haven't used it for yet>.

**Comment by estellecomment** [comment]
> We're talking about the search service here, not any particular search page

Well, make two different search services then, or different indexes for same service, or different args for same service (don't know that much about how it works sorry)

**Comment by garethbowen** [comment]
@SCdF and I have been talking about ways to make this awesomer. I quite like the deep indexing and advanced search feature and it's something we supported in 0.4 via lucene. I think it's great that right now you can configure contacts to have some arbitrary field that's specific to your deployment and search against it. There are two problems with this currently: performance (which is being covered by other issues), and relevance (this issue).

I vote for some investigation into weighted tokens. If we can make the freetext view order based on where the token is found (eg: name is a heavy weighting, parents name is a light weighting) then in this example Victor will show up at the top.

**Comment by diannakane** [comment]
We could even look at delivering the search results within labeled
sections, "names with.." "parents name with." What other types of tokens
could be returned?

We definitely need to understand more when search is used and what users
are looking for when they use it. Not immediately low-hanging fruit, in my
opinion, and should be looked at in conjunction with work on filters, which
is our other means of finding information you are looking for. Adding to
design investigations.



On Sun, May 1, 2016 at 1:43 PM, Gareth Bowen <notifications@github.com>
wrote:

> @SCdF <https://github.com/SCdF> and I have been talking about ways to
> make this awesomer. I quite like the deep indexing and advanced search
> feature and it's something we supported in 0.4 via lucene. I think it's
> great that right now you can configure contacts to have some arbitrary
> field that's specific to your deployment and search against it. There are
> two problems with this currently: performance (which is being covered by
> other issues), and relevance (this issue).
>
> I vote for some investigation into weighted tokens. If we can make the
> freetext view order based on where the token is found (eg: name is a heavy
> weighting, parents name is a light weighting) then in this example Victor
> will show up at the top.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/medic/medic-webapp/issues/2257#issuecomment-216070800>
>



-- 
P.S. Join our community. Sign up for Medic Mobile's newsletter
<http://medicmobile.us8.list-manage.com/track/click?u=e5b07d2afb4b273eaff8e3c53&id=a3129fc799&e=c511575255>
.


Dianna Kane | Senior Designer
Medic Mobile <http://www.medicmobile.org/> +1 216 262 2204
@*M*edic <http://www.twitter.com/medic> Facebook
<https://www.facebook.com/MedicMobile> @diannakane

**Comment by abbyad** [comment]
@garethbowen is this issue still relevant, or can we close it?

**Comment by garethbowen** [comment]
Yes it's still relevant as the behaviour is unchanged. There doesn't seem to be much demand to improve it though...

**Comment by michaelkohn** [comment]
With the[ search bar redesign](https://github.com/medic/cht-core/issues/7746) we did, we have been monitoring usage of the Search feature on the contacts and reports tabs and been paying extra attention to it during shadowing sessions... and in general usage is not high, but it's an area that is on our radar as we see users scrolling a lot (instead of using search) to find households on the contacts tab. I'm not convinced that it's lack of usage is due to it not working as expected or desired... I'd agree that it's often less effort (at least on a phone) to scroll through a few pages (50-120 households) of alphabetical content than tapping into a search bar, having the keyboard cover a bunch of content, typing out a few characters of what i'm looking for, hitting a button to execute the search and then look through the results set... or go through all that again to search with updated terms.  We haven't looked closely into usage of search from other areas of the system though (admin pages, messaging, etc...).

All that to say... at least from a health worker perspective, Search isn't a heavily used feature _yet_ and we (Care Teams) don't have any immediate plans to address the way it works.
